0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,264 @@
+# Object storage building block
+
+* Author(s): @ItalyPaleAle
+* State: Draft
+* Updated: 2022-12-14
+* Original proposals / discussions: dapr/dapr#4808, dapr/dapr#4934
+
+## Overview
+
+This is a design proposal for a new "object storage" building block which allows Dapr users to store and retrieve unstructured data of arbitrary size.
+
+We define **objects** as sequences of unstructured data, that should be assumed to be possibly large (many MBs or even GBs in size) and binary. Examples include images and videos, Office documents, etc.
+
+## Implementation Details
+
+### Building block interface
+
+The proposal involves creating a new `objectstore` building block with the following interface:
+
+```go
+type ObjectStore interface {
```

Comment:
What about listing objects?

Reply:
I am unsure about listing objects at this point. The problem with this is that it's a very hard API to build in a scalable, performant, and consistent way.

There could be hundreds of thousands or millions of file in a storage account / bucket, and returning them all is not practical, both for Dapr and for the service. Users *will* try to list all objects in a storage account or bucket (it's inevitable!) and that will cause problems.

The "obvious" solution here is to add pagination, which means we limit the number of objects we return in each list response, but that's actually pretty complicated in practice because some backends have no native support for pagination (local filesystems), and those who do, are inconsistent in the way it's implemented.

It's also very hard to do pagination in a consistent way (in the sense of data consistency), because if files are modified between pagination requests, the result is undefined.

My preference for now would be to move forward without listing objects, so this (already very important) proposal isn't delayed due to complications in design and implementation for supporting listing objects. It is something we could add later on, if we can figure out the correct way to do that.

Reply:
>  so this (already very important) proposal isn't delayed

Very important based on what? that's very a very subjective statement, I can't relate to that as a reviewer and any sort of urgency expressed here is redundant to the proposal.

Two things to consider here:

1. Proposals are separated from milestones. This can go anywhere from 1.11 to 1.14 and beyond.
2. Because proposals are separated from milestones, there is no requirement for every feature listed in a proposal to fit within a single milestone. Specifying the *wanted* interfaces serves as a decision record we can look back to and then derive milestone work for. How that actually works in practice remains to be seen as this is a new process for us.

I don't feel too strongly about listing objects and merely wanted to know if we view listing objects as an integral part of the overall experience.

Reply:
Alright, it wasn't clear to me that proposals did not have to be accepted and implemented in toto.

I am still unsure how to correctly implement, and even design, an API for listing; I will need to let this sit with me for a bit and perhaps even make some POCs. But I'm happy to accept suggestions here and I can update this proposal if someone has good ideas and/or experience with designing such an API.

Also, no particular urgency (in fact, the initial proposal was made in June). Just hoping the entire proposal doesn't get stalled because of lack of clarity on the listing API.

Reply:
> Just hoping the entire proposal doesn't get stalled because of lack of clarity on the listing API.

That should not happen, at least from my perspective.

Reply:
Regarding the list operation, I agree that this particular operation for most any object store is very costly and inefficient (and therefore largely unused except by individuals / tools doing administrative operations such as cleaning up stale data) -- a common solution is a keeping secondary index stored elsewhere in something that is  queryable so I think it's okay to not add to the building block.

Reply:
Listing is super important, but perhaps consider it an "optional" feature of the underlying storage? For example, local storage on disk, can list the files very quickly and cheaply, while maybe it is a burden for S3/Azure.

> The "obvious" solution here is to add pagination

To avoid paging, why not make the listing a stream that you can consume at your convenience...?